### PR TITLE
Wrap dyn type with parentheses in suggestion

### LINKED ISSUE #120223: E0310 - Invalid suggestion when type is a function with a `dyn` return type
### Code

```Rust
use std::{future::Future};

pub fn foo<T>(
    executor: impl FnOnce(T) -> dyn Future<Output = ()>,
) -> Box<dyn FnOnce(T) -> dyn Future<Output = ()>> {
    Box::new(executor)
}
```

### Current output

```Shell
error[E0310]: the parameter type `impl FnOnce(T) -> dyn Future<Output = ()>` may not live long enough
 --> src/lib.rs:6:5
  |
6 |     Box::new(executor)
  |     ^^^^^^^^^^^^^^^^^^
  |     |
  |     the parameter type `impl FnOnce(T) -> dyn Future<Output = ()>` must be valid for the static lifetime...
  |     ...so that the type `impl FnOnce(T) -> dyn Future<Output = ()>` will meet its required lifetime bounds
  |
help: consider adding an explicit lifetime bound
  |
4 |     executor: impl FnOnce(T) -> dyn Future<Output = ()> + 'static,
  |                                                         +++++++++
```

### Desired output

```Shell
error[E0310]: the parameter type `impl FnOnce(T) -> dyn Future<Output = ()>` may not live long enough
 --> src/lib.rs:6:5
  |
6 |     Box::new(executor)
  |     ^^^^^^^^^^^^^^^^^^
  |     |
  |     the parameter type `impl FnOnce(T) -> dyn Future<Output = ()>` must be valid for the static lifetime...
  |     ...so that the type `impl FnOnce(T) -> dyn Future<Output = ()>` will meet its required lifetime bounds
  |
help: consider adding an explicit lifetime bound
  |
4 |     executor: impl FnOnce(T) -> (dyn Future<Output = ()>) + 'static,
  |                                 +                       +++++++++++
```

### Rationale and extra context

The issue is that the diagnostic suggests adding `+ 'static`
However performing this modification exactly as specified results in a syntax error:
```
error: ambiguous `+` in a type
 --> src/lib.rs:4:33
  |
4 |     executor: impl FnOnce(T) -> dyn Future<Output = ()> + 'static,
  |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use parentheses to disambiguate: `(dyn Future<Output = ()> + 'static)`                                                     ^^^^^^^^^^^^ expected a path
```

And then if you follow this syntax error's suggestion, you get back to the original state:

```
error[E0310]: the parameter type `impl FnOnce(T) -> (dyn Future<Output = ()> + 'static)` may not live long enough
 --> src/lib.rs:6:5
  |
6 |     Box::new(executor)
  |     ^^^^^^^^^^^^^^^^^^
  |     |
  |     the parameter type `impl FnOnce(T) -> (dyn Future<Output = ()> + 'static)` must be valid for the static lifetime...
  |     ...so that the type `impl FnOnce(T) -> (dyn Future<Output = ()> + 'static)` will meet its required lifetime bounds
  |
help: consider adding an explicit lifetime bound
  |
4 |     executor: impl FnOnce(T) -> (dyn Future<Output = ()> + 'static) + 'static,
  |                                                                     +++++++++

For more information about this error, try `rustc --explain E0310`.
```

However if instead it suggested wrapping the return value in `()` and then adding `+ static` - you'd get to the right state immediately (I think? I'm  still a rust newbie)

### Other cases

_No response_

### Rust Version

```Shell
Stable: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=233f12d3bc7a1958232137e4b27e2921
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by long-long-float**
Close #120223

Fix wrong suggestion that is grammatically incorrect.
Specifically, I added parentheses to dyn types that need lifetime bound.

```
help: consider adding an explicit lifetime bound
  |
4 |     executor: impl FnOnce(T) -> (dyn Future<Output = ()>) + 'static,
  |                                 +                       +++++++++++
```

### CODE REVIEW

**Comment by fmease** [comment]
Sure, I plan on reviewing this later today. Sorry for taking a bit, I'm assigned to several PRs and continuously work on PRs myself ^^'

**Review by fmease — CHANGES_REQUESTED** [review]
I'm sorry for leaving you hanging like that! I have a couple of suggestions for how to structure this a bit differently and some correctness things.

At the moment your changes only trigger in very few cases. I think we should generalize it to apply to more `+ XXX`-style suggestions.

First and foremost, I would check the other 3 usage sites `bounds_span_for_suggestion` whether they could benefit from this, too (I even suggest you to merge `bounds_span_for_suggestion_with_parentheses` into it in one of my review comments). You could even consider `grep`'ing through `tests/ui/` to check if there are any other incorrect *consider adding an explicit lifetime bound* suggestions.

In summary, I'm a fan of fixing the incorrect diagnostic suggestion but I'd like to make sure to tap the full potential.

**Comment by long-long-float** [comment]
@fmease Thank you for detailed review! I will fix them and check for more patterns in a couple of days.

**Comment by long-long-float** [comment]
I will fix testing failure at first.

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -647,6 +662,35 @@ impl<'hir> Generics<'hir> {
...
+    /// Returns bounds span for suggestions.
+    /// If the span including lifetime bound needs parentheses, it returns a tuple of a span to be surrounded by parentheses and true.
+    /// e.g. `dyn Future<Output = ()> + 'static` needs parentheses `(dyn Future<Output = ()>) + 'static`
+    pub fn bounds_span_for_suggestions_with_parentheses(
+        &self,
+        param_def_id: LocalDefId,
+    ) -> Option<(Span, bool)> {
+        self.bounds_for_param(param_def_id).flat_map(|bp| bp.bounds.iter().rev()).find_map(
+            |bound| {
+                let span_for_parentheses = bound.get_inner_ty().and_then(|ty| {
```
[the remark this answers is not in the document]
Comment (long-long-float) [reply compiler/rustc_hir/src/hir.rs]: I tried to call `tcx.is_fn_trait` but `Generics` doesn't have `tcx`.
I also tried to pass `tcx` via an argument but `rustc_hir` doesn't have reference to `rustc_middle` which has `TyCtxt`.
Can I add a reference to `rustc_middle` or is there more good solution?

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -446,6 +446,21 @@ impl GenericBound<'_> {
         }
     }
 
+    pub fn get_inner_ty(&self) -> Option<&Ty<'_>> {
+        match self {
+            GenericBound::Trait(data, _) => {
+                let segment = data.trait_ref.path.segments.first()?;
```
[the remark this answers is not in the document]
Comment (long-long-float) [reply compiler/rustc_hir/src/hir.rs]: This will be solved by your suggested code above.

File: compiler/rustc_hir/src/hir.rs [hunk 3]
```diff
@@ -647,6 +662,35 @@ impl<'hir> Generics<'hir> {
         )
     }
 
+    /// Returns bounds span for suggestions.
+    /// If the span including lifetime bound needs parentheses, it returns a tuple of a span to be surrounded by parentheses and true.
+    /// e.g. `dyn Future<Output = ()> + 'static` needs parentheses `(dyn Future<Output = ()>) + 'static`
+    pub fn bounds_span_for_suggestions_with_parentheses(
```
[the remark this answers is not in the document]
Comment (long-long-float) [reply compiler/rustc_hir/src/hir.rs]: Completed at https://github.com/rust-lang/rust/pull/120929/commits/a0452d36131e1cf52656bd366f2499cb57cafa15

File: compiler/rustc_hir/src/hir.rs [hunk 4]
```diff
@@ -647,6 +662,35 @@ impl<'hir> Generics<'hir> {
...
+                        None
+                    }
+                });
+
+                span_for_parentheses.map_or_else(
+                    || {
+                        let bs = bound.span();
+                        bs.can_be_used_for_suggestions().then(|| (bs.shrink_to_hi(), false))
+                    },
+                    |span| Some((span, true)),
```
[the remark this answers is not in the document]
Comment (long-long-float) [reply compiler/rustc_hir/src/hir.rs]: Completed at https://github.com/rust-lang/rust/pull/120929/commits/a0452d36131e1cf52656bd366f2499cb57cafa15

**Comment by long-long-float** [comment]
@fmease 

> First and foremost, I would check the other 3 usage sites `bounds_span_for_suggestion` whether they could benefit from this, too.

I checked them:

* `compiler/rustc_hir_typeck/src/method/suggest.rs`
	* `tests/ui/suggestions/impl-trait-with-missing-trait-bounds-in-arg.rs`
		* The impl Trait accepts `impl Foo + Bar` so we don't need parenthesis.
		* I think this suggestion is for type parameters (impl Trait in arguments), not `dyn`.
* `compiler/rustc_middle/src/ty/diagnostics.rs`
	* Same as above?
* `compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs`
	* `?Sized` cannot be used in trait object types (`dyn`).

> You could even consider grep'ing through `tests/ui/` to check if there are any other incorrect consider adding an explicit lifetime bound suggestions.

I searched `consider adding an explicit lifetime bound` under `tests/ui`  and couldn't find any incorrect suggestions.

I think incorrect suggestions which need parentheses happen with only `dyn` (I'm not familiar with Rust, so it may incorrect).

File: compiler/rustc_hir/src/hir.rs [hunk 5]
```diff
@@ -644,13 +644,53 @@ impl<'hir> Generics<'hir> {
...
-                bs.can_be_used_for_suggestions().then(|| bs.shrink_to_hi())
+                let span_for_parentheses = get_inner_ty(bound).and_then(|ty| {
+                    if let TyKind::TraitObject(
+                        _,
+                        _,
+                        TraitObjectSyntax::Dyn | TraitObjectSyntax::DynStar,
+                    ) = ty.kind
+                    {
+                        let span = ty.span;
+                        span.can_be_used_for_suggestions().then(|| span)
```
Comment (fmease) [inline compiler/rustc_hir/src/hir.rs:679]:
```suggestion
                        ty.span.can_be_used_for_suggestions().then_some(ty.span)
```

File: compiler/rustc_hir/src/hir.rs [hunk 6]
```diff
@@ -644,13 +644,53 @@ impl<'hir> Generics<'hir> {
...
-                let bs = bound.span();
-                bs.can_be_used_for_suggestions().then(|| bs.shrink_to_hi())
+                let span_for_parentheses = get_inner_ty(bound).and_then(|ty| {
+                    if let TyKind::TraitObject(
+                        _,
+                        _,
+                        TraitObjectSyntax::Dyn | TraitObjectSyntax::DynStar,
+                    ) = ty.kind
+                    {
+                        let span = ty.span;
```
Comment (fmease) [inline compiler/rustc_hir/src/hir.rs:678]:
```suggestion
                        ty.span.can_be_used_for_suggestions().then_some(ty.span)
```

File: compiler/rustc_hir/src/hir.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (fmease) [reply compiler/rustc_hir/src/hir.rs]: Oops, I thought we were in `rustc_middle` ^^' In this case we should just check if the args are parenthesized via `segment.args().parenthesized == GenericArgsParentheses::ParenSugar` which is more general (e.g., it also accounts for `async Fn()` etc.).

[the remark this answers is not in the document]
Comment (long-long-float) [reply compiler/rustc_hir/src/hir.rs]: > In this case we should just check if the args are parenthesized via `segment.args().parenthesized == GenericArgsParentheses::ParenSugar` which is more general (e.g., it also accounts for async Fn() etc.).

Sorry, I couldn't understand why is this condition needed.
My understanding is that it is needed to avoid `async Fn() + 'static`, is it right?
I would appreciate if you could give me an example to use this.

[the remark this answers is not in the document]
Comment (fmease) [reply compiler/rustc_hir/src/hir.rs]: The check `segment.args().parenthesized == ParenSugar` should prevent us from adding unnecessary (or even incorrect? dunno) parentheses in cases like `impl SomeTrait<Output = dyn Future<Output = ()>>`.

[the remark this answers is not in the document]
Comment (long-long-float) [reply compiler/rustc_hir/src/hir.rs]: Thank you! I understood and fixed.

File: compiler/rustc_hir/src/hir.rs [hunk 7]
```diff
@@ -644,13 +644,41 @@ impl<'hir> Generics<'hir> {
...
             |bound| {
-                // We include bounds that come from a `#[derive(_)]` but point at the user's code,
-                // as we use this method to get a span appropriate for suggestions.
-                let bs = bound.span();
-                bs.can_be_used_for_suggestions().then(|| bs.shrink_to_hi())
+                let span_for_parentheses = if let Some(trait_ref) = bound.trait_ref()
+                    && let [.., segment] = trait_ref.path.segments
+                    && segment.args().parenthesized == GenericArgsParentheses::ParenSugar
+                    && let [binding] = segment.args().bindings
+                    && let TypeBindingKind::Equality { term: Term::Ty(ret_ty) } = binding.kind
```
Comment (fmease) [inline compiler/rustc_hir/src/hir.rs:660]: Ah, I think you need to [`peel_refs`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Ty.html#method.peel_refs) here, so that we detect things like `Fn() -> &'static dyn Future<Output = ()>` (see my other review comments for examples).

```suggestion
                    && let TypeBindingKind::Equality { term: Term::Ty(ret_ty) } = binding.kind
                    && let ret_ty = ret_ty.peel_refs()
```

**Review by fmease — CHANGES_REQUESTED** [review]
Thanks a lot!

> First and foremost, I would check the other 3 usage sites `bounds_span_for_suggestion` whether they could benefit from this, too.

I could find an example for each usage site. Please make them use `open_paren_sp` as well, thanks! :)

After that, this PR should be good to go! One more thing, could you squash the commits into one (or fewer) commits?

**Comment by long-long-float** [comment]
@fmease 
Thank you for reviewing! I fixed to use `open_paren_sp` in all sites, added tests, and squashed commits. Could you check them?

**Review by fmease — APPROVED** [review]

**Comment by fmease** [comment]
> https://github.com/rust-lang/rust/pull/120929#issuecomment-2067872794

CI still has errors, please fix them. If you have any questions regarding the failures, don't hesitate to reach out and ask them!

**Comment by long-long-float** [comment]
@fmease Thank you! All tests are passed.

**Review by fmease — APPROVED** [review]
